rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf

```diff
@@ -1164,6 +1164,84 @@ SecRule REQUEST_HEADERS:Accept-Encoding "@gt 50" \
     severity:'CRITICAL',\
     setvar:'tx.inbound_anomaly_score_pl1=+%{tx.critical_anomaly_score}'"
 
+#
+# Restrict response charsets that we allow.
+# The following rules make sure that the response will be in an ASCII-compatible charset that
+# phase 4 rules can properly understand and block.
+#
+
+#
+# Some servers rely on the request Accept header to determine what charset to respond with.
+# This rule restricts these to familiar charsets.
+#
+# RegEx generated using:
+# python3 util/regexp-tricks/negative-lookahead.py --prefix ";\s*charset\s*=\s*" utf-8 iso-8859-1 --suffix "[^;,]*"
+#
+SecRule REQUEST_HEADERS:Accept "@rx ;\s*charset\s*=\s*(?:[^ui]|u[^t]|ut[^f]|utf[^-]|utf-[^8]|i[^s]|is[^o]|iso[^-]|iso-[^8]|iso-8[^8]|iso-88[^5]|iso-885[^9]|iso-8859[^-]|iso-8859-[^1])[^;,]*" \
+    "id:920600,\
+    phase:1,\
+    block,\
+    t:none,t:lowercase,\
+    msg:'Illegal Accept header: charset parameter',\
+    logdata:'%{MATCHED_VAR}',\
+    tag:'application-multi',\
+    tag:'language-multi',\
+    tag:'platform-multi',\
+    tag:'attack-protocol',\
+    tag:'paranoia-level/1',\
+    tag:'OWASP_CRS',\
+    ver:'OWASP_CRS/4.0.0-rc1',\
+    severity:'CRITICAL',\
+    setvar:'tx.inbound_anomaly_score_pl1=+%{tx.critical_anomaly_score}'"
+
+
+#
+# Accept-Charset is a forbidden header that should not be used by clients and ignored by servers.
+# Completely block it if it's found.
+#
+# Reference: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-Charset
+#
+SecRule REQUEST_HEADERS:Accept-Charset "@unconditionalMatch" \
+    "id:920610,\
+    phase:1,\
+    block,\
+    t:none,\
+    msg:'Illegal header: Accept-Charset',\
+    logdata:'%{MATCHED_VAR}',\
+    tag:'application-multi',\
+    tag:'language-multi',\
+    tag:'platform-multi',\
+    tag:'attack-protocol',\
+    tag:'paranoia-level/1',\
+    tag:'OWASP_CRS',\
+    ver:'OWASP_CRS/4.0.0-rc1',\
+    severity:'CRITICAL',\
+    setvar:'tx.inbound_anomaly_score_pl1=+%{tx.critical_anomaly_score}'"
+
+
+#
+# The Content-Encoding header is used to list any encodings that have been applied to the original payload.
+# It is only used for compression, which isn't supported by CRS by default since it blocks newlines and null bytes inside the request body.
+# Most compression algorithms require at least null bytes per RFC. Blocking it shouldn't break anything and increases security
+# since ModSecurity is incapable of properly scanning compressed request bodies.
+#
+SecRule REQUEST_HEADERS:Content-Encoding "@unconditionalMatch" \
```

Comment:
Is this safe to add on Paranoia Level 1? I could see this cause a lot of false positives.

Reply:
@studersi, yes, that is correct :)